rfcs/table-destructuring.md

```diff
@@ -0,0 +1,236 @@
+```lua
+-- Example using a RemoteFunction callback
+
+RemoteFunction.InvokeServer = function(player, { target, damage })
+    print("Dealt " .. damage .. " to " .. target.Name)
+end
+```
+
+## Design
+
+### Variable Assignments for Arrays
+With the ability to use `unpack` to destructure arrays, the need for this feature can seem negligible. Nevertheless, `unpack` cannot be used on dictionaries.
+```lua
+-- Current
+local foo = {1, 2, 3}
+
+local one, two, three = unpack(foo) -- 1, 2, 3
+```
+```lua
+-- New
+local foo = {1, 2, 3}
+
+local {one, two, three} = foo -- 1, 2, 3
+```
+
+### Variable Assignments for Dictionaries
+```lua
+-- Current
+local foo = {
+    red = "good";
+    green = "better";
+    blue = "best";
+}
+
+local red, green, blue = foo.red, foo.green, foo.blue
+print(red) -- good
+print(green) -- better
+print(blue) -- best
+```
+```lua
+-- New
+local foo = {
+    red = "good";
+    green = "better";
+    blue = "best";
+}
+
+local {red, green, blue} = foo
+print(red) -- good
+print(green) -- better
+print(blue) -- best
+```
+
+### Variable Assignments for Function Returns
+```lua
+-- Current
+local function f()
+    return {
+        name = "John";
+        age = 25;
+        gender = "male";
+    }
+end
+
+local data = f()
+local name, age, gender = data.name, data.age, data.gender
+
+print(name) -- John
+print(age) -- 25
+print(gender) -- male
+```
+```lua
+-- New
+local function f()
+    return {
+        name = "John";
+        age = 25;
+        gender = "male";
+    }
+end
+
+local {name, age, gender} = f()
+print(name) -- John
+print(age) -- 25
+print(gender) -- male
+```
+
+### Variable Assignments for UserData & Vectors
+```lua
+-- Current
+local Part = workspace.Part
+local Position = Part.CFrame.Position
+local X, Y, Z = Position.X, Position.Y, Position.Z
+
+print(X, Y, Z) -- 0, 0, 0
+```
+```lua
+-- New
+local Part = workspace.Part
+local {Position} = Part.CFrame
+local {X, Y, Z} = Position
+
+print(X, Y, Z) -- 0, 0, 0
+```
+
+### Variable Assignments for Indexing Children (???)
+```lua
+-- Current
+local Baseplate = workspace.Baseplate
+print(Baseplate) -- Baseplate
+```
+```lua
+-- New
+local {Baseplate} = workspace
+print(Baseplate) -- Baseplate
+```
+
+---
+
+### Destructuring Function Parameters
+```lua
+-- Before
+local function f(data)
+    local name = data.name
+    local age = data.age
+    local gender = data.gender
+
+    print(name, age, gender) -- John, 25, male
+end
+
+f({
+    name = "John";
+    age = 25;
+    gender = "male";
+})
+```
+```lua
+-- New
+local function f({ name, age, gender })
+    print(name, age, gender) -- John, 25, male
+end
+
+f({
+    name = "John";
+    age = 25;
+    gender = "male";
+})
+```
+
+---
+
+### Defining Alternative Identifiers (Questionable?)
+In this example, `fizz-buzz` is not a valid identifier. You should be able to instead define an alternative identifier, like so.
+```lua
+local foo = {
+    ["fizz-buzz"] = true;
+}
+
+local { fizzbuzz = "fizz-buzz" } = foo
```

Comment:
Why not use a new keyword `as` instead? It would be more at home with Luau's syntax.

```lua
local { ["fizz-buzz"] as fizzbuzz } = foo
```

It would also make this possible:
```lua
local { [1] as first } = { 1 }
```